pkg/policy/mapstate.go

```diff
@@ -114,55 +114,91 @@ func (e *MapStateEntry) IsRedirectEntry() bool {
-		// the new deny takes precedence over all L3-L4 for each the L3 is the
-		// same on both the existing rules and new deny rule.
-		if key.DestPort == 0 {
+func (keys MapState) DenyPreferredInsert(newKey Key, newEntry MapStateEntry) {
+	if newEntry.IsDeny {
+		// case for an existing allow L4-only and we are inserting deny L3-only
+		if newKey.DestPort == 0 && newKey.Nexthdr == 0 && newKey.Identity != 0 {
+			l4OnlyAllows := MapState{}
+			for k, v := range keys {
+				if newKey.TrafficDirection == k.TrafficDirection &&
+					!v.IsDeny &&
+					k.Identity == 0 {
+					// create a deny L3-L4 with the same allowed L4 port and proto
+					newKeyCpy := newKey
+					newKeyCpy.DestPort = k.DestPort
+					newKeyCpy.Nexthdr = k.Nexthdr
+					keys[newKeyCpy] = newEntry
+
+					l4OnlyAllows[k] = v
+				}
+			}
+			// Delete all L3-L4 if we are inserting a deny L3-only and
+			// there aren't allow L4-only for the existing deny L3-L4
 			for k := range keys {
-				if k.TrafficDirection == key.TrafficDirection &&
-					k.Identity == key.Identity {
-					delete(keys, k)
+				if k.TrafficDirection == newKey.TrafficDirection &&
+					k.DestPort != 0 && k.Nexthdr != 0 &&
+					k.Identity == newKey.Identity {
+
+					kCpy := k
+					kCpy.Identity = 0
+					if _, ok := l4OnlyAllows[kCpy]; !ok {
+						delete(keys, k)
+					}
 				}
 			}
+		} else {
+			// If there is a Deny that is an L3-only, all L3-L4 denies won't be
+			// accepted because a L3-Deny takes precedence.
+			newKeyCpy := newKey
+			newKeyCpy.DestPort = 0
+			newKeyCpy.Nexthdr = 0
+			v, ok := keys[newKeyCpy]
+			if ok && v.IsDeny {
+				// Found a L3-only Deny so we won't accept any L3-L4 policies
+				return
+			}
 		}
-		keys[key] = entry
-		return
-	}
 
-	if _, ok := keys[key]; ok {
-		// Key already exist, keep the existing entry so that
-		// a deny entry is never overwritten by an allow entry
+		keys[newKey] = newEntry
+		return
+	} else if newKey.Identity == 0 {
+		// case for an existing deny L3-only and we are inserting allow L4-only
+		for k, v := range keys {
+			if newKey.TrafficDirection == k.TrafficDirection {
+				if v.IsDeny && k.Identity != 0 && k.DestPort == 0 && k.Nexthdr == 0 {
+					// create a deny L3-L4 with the same deny L3
+					newKeyCpy := newKey
+					newKeyCpy.Identity = k.Identity
+					keys[newKeyCpy] = v
+				}
+			}
+		}
+		keys[newKey] = newEntry
 		return
 	}
-	// If there is a Deny that is an L3-only, all L3-L4 allows won't be
-	// accepted because a L3-Deny takes precedence.
-	oldPort := key.DestPort
-	oldNexthdr := key.Nexthdr
-	key.DestPort = 0
-	key.Nexthdr = 0
-	v, ok := keys[key]
+	// branch for adding a new allow L3-L4
+	newKeyCpy := newKey
+	newKeyCpy.DestPort = 0
+	newKeyCpy.Nexthdr = 0
+	v, ok := keys[newKeyCpy]
 	if ok && v.IsDeny {
-		// Found a L3-only Deny so we won't accept any L3-L4 policies
+		// Found a L3-only Deny so we won't accept any L3-L4 allow policies
 		return
 	}
```

Comment:
What's different between these lines below and lines 162-171? Oh, I guess whether the new entry is a deny or an allow entry. Hm. Possible opportunity for code sharing?

Reply:
It is tricky to share the code across both branches. I'm open to suggestions!